kairos-server/src/l1_sync/event_manager.rs

```diff
@@ -0,0 +1,89 @@
+use std::sync::Arc;
+
+use casper_event_toolkit::fetcher::{Fetcher, Schemas};
+use casper_event_toolkit::metadata::CesMetadataRef;
+use casper_event_toolkit::rpc::client::CasperClient;
+
+use crate::state::ServerStateInner;
+use kairos_circuit_logic::transactions::{KairosTransaction, L1Deposit};
+
+use super::error::L1SyncError;
+
+pub struct EventManager {
+    next_event_id: u32,
+    fetcher: Option<Fetcher>,
+    schemas: Option<Schemas>,
+    server_state: Arc<ServerStateInner>,
+}
+
+impl EventManager {
+    pub fn new(server_state: Arc<ServerStateInner>) -> Self {
+        EventManager {
+            next_event_id: 0,
+            fetcher: None,
+            schemas: None,
+            server_state,
+        }
+    }
```

Comment:
I don't see where it makes sense to model the `fetcher` and `schemas` attributes as optionals and introduce a new type that basically sets up the user for using the api in a wrokng way, and moreover forces a user to have a two-step setup process ending up with these two possible use scenarios:
(Notice that the case `fetcher == Some && schema == None` and `fetcher == None && schema Some` is not even possible)
```rust
// correct
let event_manager = EventManager::new(); // fetcher == None && schema == None
event_manager.initialize(); // fetcher == Some(..) && schema == Some(..)
event_manager.process_new_events();

// wrong usage, but possible without any compiler warning
let event_manager = EventManager::new();  // fetcher == None && schema == None
event_manager.process_new_events(); // This will fail and I think I should not even be able to call this and a strong indicator is a missing fetcher and schema, which can both be used to prevent an eventuall call

// better: process_new_events is now just a function in the server crate and can belong semantically to the server
// the ServerState contains all the information like the `rpc_url`, `contract_hash`, `server_state`, even the `next_event_id`
pub fn process_new_events(server_state: &ServerState, callback: FnOnce -> Result) -> Result {
// initialize the fethcer. Could fail, hence result return type
// initialize the schema. Could fail, hence result return type
// call callback which injects any behavior a user might want to use which could potentially fail, hence Result return type of the callback
}
```
The latter being a single function call, which deals with all the details a user does not care about. The user only cares about what should happen when new events are observed, reducing the amount of knowledge to just implement the callback logic.

Reply:
I removed partial construction pattern, so `fetcher` and `schemas` are no longer optional: https://github.com/cspr-rad/kairos/pull/96/commits/50a3654c33e72267c89df53df61ca72a2bcca9eb.